src/Equinox.Core/Caching.fs

```diff
@@ -3,66 +3,59 @@ module Equinox.Core.Caching
 open Equinox.Core.Tracing
 open Serilog
 open System
-open System.Diagnostics
 open System.Threading
 open System.Threading.Tasks
 
-type IReloadableCategory<'event, 'state, 'context> =
-    inherit ICategory<'event, 'state, 'context>
-
+type IReloadable<'state> =
     abstract Reload: log: ILogger * streamName: string * requireLeader: bool * streamToken: StreamToken * state: 'state * ct: CancellationToken
                      -> Task<struct (StreamToken * 'state)>
 
-type private Decorator<'event, 'state, 'context>
-    (   category: IReloadableCategory<'event, 'state, 'context>,
-        tryReadCache: string -> Task<voption<struct (StreamToken * 'state)>>,
-        updateCache: string -> struct (StreamToken * 'state) -> Task<unit>) =
-    let cache streamName (inner: CancellationToken -> Task<struct (StreamToken * 'state)>) ct = task {
-        let! tokenAndState = inner ct
-        do! updateCache streamName tokenAndState
-        return tokenAndState }
+let private tee f (inner: CancellationToken -> Task<struct (StreamToken * 'state)>) ct = task {
+    let! tokenAndState = inner ct
+    do! f tokenAndState
+    return tokenAndState }
+
+type private Decorator<'event, 'state, 'context, 'cat when 'cat :> ICategory<'event, 'state, 'context> and 'cat :> IReloadable<'state> >
+    (category: 'cat, cache: ICache, compareTokens, createKey, createOptions) =
+    let tryRead key = task {
+        let! cacheItem = cache.TryGet key
+        let act = System.Diagnostics.Activity.Current
+        if act <> null then act.AddCacheHit(ValueOption.isSome cacheItem) |> ignore
+        return cacheItem }
+    let save key (inner: CancellationToken -> Task<struct (StreamToken * 'state)>) ct = task {
+        let! struct (token, state) as res = inner ct
+        do! cache.UpdateIfNewer(key, compareTokens, createOptions (), token, state)
+        return res }
     interface ICategory<'event, 'state, 'context> with
-        member _.Load(log, categoryName, streamId, streamName, allowStale, requireLeader, ct) = task {
-            match! tryReadCache streamName with
```

Comment:
note this was not computing a correct key in the case of SlidingWindowPrefixed (fixed in #386)